lib/generators/administrate/routes/routes_generator.rb

```diff
@@ -50,30 +53,92 @@ def dashboard_resources
         end
       end
 
+      def dashboard_resources_hash
+        output = [{}, []]
+        dashboard_resources.each do |resource_path|
+          split_path = resource_path.split("/")
+          output = [
+            add_namespace_to_hash(split_path, output[0]),
+            add_resources_to_hash(split_path, output[1]),
+          ]
+        end
+        output
+      end
```

Comment:
I see what you are trying to do there. It's something complicated to code and test, so I can't complain about your implementation. What I'm concerned is that, precisely because it's difficult to test, it probably doesn't cover as many cases as we think.

Here are a couple (asking without actually checking, perhaps you have this covered). Does it support...:

* Multiple levels of nesting?
* Resources with nesting? Example:

```ruby
resources :authors do # NOTE: not `namespace`
  resources :posts
end
# This gives you things like:
#  - /authors
#  - /authors/:id
#  - /authors/:id/posts
#  - /authors/:id/posts/:id
```

Testing this is difficult with our current spec, which is a bit more end-to-end testing the functionality of the generator. If we wanted to test these edge cases (which I think we should), we would end up complicating that spec too much.

Instead, I think this "figuring out the tree" functionality should be moved to a separate class that can be unit-tested for edge cases, leaving a general example for this spec. It could have an interface like this:

```ruby
resources = [["blog/authors"], ["blog/posts"], ["countries"], ["customers"]]
rtc = RouteTreeCalculator.new(resources) # First name that came to mind; not attached to it
rtc.tree # => Not sure exactly what, but something that `generate_resource_routes` will consume
```

Does this make sense?

Reply:
The hash output works well (I wrote it in a separate folder along with some tests) I think moving the function out into new folder in administrate will make the testing and code clearer.

I am confident it leads to the correct nesting structure - though I labeled them all as a namespace instead of a resource so that would need to change? or we need to figure out when to use namespace and not resource..

Reply:
moved into a new helper file - should make putting tests for it into a sensible location (not recreated yet)

Reply:
`resources :authors do # NOTE: not `namespace`
  resources :posts
end
# This gives you things like:
#  - /authors
#  - /authors/:id
#  - /authors/:id/posts
#  - /authors/:id/posts/:id`
 
 Only just looking back at this - I am pretty confident that what is currently in place will function but leads to 
 `resources :authors do
  resources :posts
end
resources: authors`

I feel like this is something to think about next weekend